eoscompanion/routes.py

```diff
@@ -161,19 +160,419 @@ def desktop_id_to_app_id(desktop_id):
+    entry.
+
+    Problems with individual models will be reported on the console.
+    '''
+    # A map of sources, first by type and then by a unique identifier
+    # for the source of that type. This is used to quickly lookup whether
+    # an entry has a corresponding source.
+    sources_map = defaultdict(set)
+
+    for source in sources:
+        sources_map[source['type']].add(get_source_identifier(source))
+
+    for model in models:
+        # Get the app ID for all content models that have a desktop ID. If
+        # a content model does not have a desktop ID then we can still
+        # display it, but the user won't be able to navigate to content
+        # for that app.
+        try:
+            app_id = desktop_id_to_app_id(model.get_property('desktop-id'))
+        except TypeError:
+            app_id = None
+
+        if app_id is not None and app_id not in sources_map['application']:
+            logging.warning('Model %s with app-id %s did not have a '
+                            'corresponding entry in the sources list, it will be '
+                            'ignored', model, app_id)
+            continue
+
+        for entry in content_feed_model_to_json_entries(model, app_id, query):
+            yield entry
+
+
 @require_query_string_param('deviceUUID')
 @require_query_string_param('mode')
 @record_metric('af3e89b2-8293-4703-809c-8e0231c128cb')
-def companion_app_server_feed_route(server, msg, path, query, *args):
-    '''Send feed.'''
-    del server
+def companion_app_server_feed_route(server,
+                                    msg,
+                                    path,
+                                    query,
+                                    context,
+                                    content_db_conn):
+    '''Request the Content Feed from EosDiscoveryFeed.
+
+    The :mode: paramter is used to determine whether newer entries
+    should be fetched or older entries should be fetched. It is currently
+    unused for now, the feed is the same as the Discovery Feed on the
+    desktop for now.
+    '''
     del path
-    del args
+    del context
+
+    def _on_received_ordered_feed_models(error, models):
+        '''Callback for when the ordered feed models are ready.
+
+        The models are in the exact order that they should be displayed
+        in the feed, so we should marshal them into the appropriate
+        JSON representation and display them now.
+        '''
+        def _on_received_sources(error, sources):
+            '''Callback for when we work out the info for all the sources.'''
+            if error is not None:
+                json_response(msg, {
+                    'status': 'error',
+                    'error': serialize_error_as_json_object(
+                        EosCompanionAppService.error_quark(),
+                        EosCompanionAppService.Error.FAILED,
+                        detail={
+                            'server_error': str(error)
```

Comment:
Did we not say to use `message` for these?

Reply:
We did! Legacy strikes again